### DISCUSSION TITLE
Detect Eclipse version when setting up JUnit classpath?

### DISCUSSION THREAD

**Discussion by lppedd** [eclipse-tycho/tycho#1760] [General]
I was looking at this change:

![image](https://user-images.githubusercontent.com/19871649/204619049-6f94d188-b387-46dc-98c6-5d27cbcf9334.png)

And TBH it looks too forced. Like, this means I have to adjust the target platform as @Bananeweizen mentioned in another discussion.
Do you think there is a way to detect the environment and adjust accordingly?

I couldn't even find a separate p2 repository for JUnit 5, and as of now I wouldn't like to ditch the repository-based target platform.

**Comment by laeubi** [comment]
There are always ways but everything comes at a price, that is development effort and maintenance efforts. Beside that, there are releases and if one want to depend on the old stuff one can use old releases as well but this should not hinder those that want to move forward.

> I couldn't even find a separate p2 repository for JUnit 5

You don't need a separate one, you can simply add one that contains this and pull in individual units. Or even just upgrade your products and benefit from many bugfixes and enhancements.

That said, if you like to propose a patch with an integration test that covers both cases and it does not adds too much complications to the code it could be considered.

**Reply by lppedd** [reply]
@laeubi 

> if you like to propose a patch

I definitely would. However I currently have no idea where to start looking, being it something related to the Eclipse environment. If you have a couple pointers, I can look by myself after. Thanks!

**Reply by lppedd** [reply]
Maybe an easy solution could be to look for the new ones, and revert to the old ones if needed? However this happens during dependency resolution and I don't know how to catch the resolution error and "try again".

**Reply by laeubi** [reply]
There is no "Eclipse" environment, and actually there is no second chance for resolving ... 

The very first step would be to provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue, then one might look at https://github.com/eclipse-tycho/tycho/blob/f2f037155575c31cc8eab71fc7d74e3a8d175ba2/tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/OsgiBundleProject.java#L370-L384

this is where classpath resolving is happen, an maybe one can have a method `Collection<ArtifactKey> getAlternatingArtifacts()` in `JUnitClasspathContainerEntry` to allow checking another set of artifacts if the first one fails. This would then be the part that should make the test pass the compilation phase.

Then the following part come into play:
https://github.com/eclipse-tycho/tycho/blob/f2f037155575c31cc8eab71fc7d74e3a8d175ba2/tycho-core/src/main/java/org/eclipse/tycho/core/osgitools/OsgiBundleProject.java#L841-L849

where you probably want to have stored a flag in the reactor project in the previous step if the first or the second set of artifacts was successfully resolved and act accordingly.

**Reply by lppedd** [reply]
@laeubi nice! I had a look, but found there is another point to consider.

https://github.com/eclipse-tycho/tycho/blob/66e864fecb7a3aa78c4b4e34749d3aecb491115e/tycho-core/src/main/java/org/eclipse/tycho/core/resolver/DefaultTychoResolver.java#L156-L175

This fails the entire build before arriving to `OsgiBundleProject`.

**Reply by laeubi** [reply]
I didn't said its trivial ;-)

But `OSGiBundleProject` actually is  `BundleProject` ... and it calls the method mentioned above.

**Reply by lppedd** [reply]
@laeubi mmh, you're right, however it seems `OsgiBundleProject#computeExtraTestClasspath` is not invoked when building the target platform in `DefaultTychoResolver`. 

In fact I think that `computeExtraTestClasspath` method actually requires an already built TP.  
Basically `getExtraTestRequirements` is called earlier than `computeExtraTestClasspath`.

**Reply by lppedd** [reply]
I think I would probably need access to the preliminary target platform. But it seems its access is locked.

**Reply by lppedd** [reply]
@laeubi so to update you on this. I've opened up the preliminary target platform and got it working.
The reasoning is "do we have at least one JUnit 5 post-name-change dependency?".
If yes, use the new names, otherwise revert back to the old ones.
A pity I've found the blocking `JupiterTestEngine` issue.
